dex/calc/fees.go

```diff
@@ -1,14 +1,29 @@
 package calc
 
-import "decred.org/dcrdex/dex"
-
-// RequiredFunds calculates the minimum amount needed to fulfill the swap amount
-// and pay transaction fees. The backingSize is the sum size of the serialized
-// inputs associated with a set of UTXOs to be spent. The swapVal is the total
-// quantity needed to fulfill an order.
-func RequiredFunds(swapVal uint64, backingSize uint32, coin *dex.Asset) uint64 {
-	R := float64(coin.SwapSize) * float64(coin.FeeRate) / float64(coin.LotSize)
-	fBase := uint64(float64(swapVal) * R)
-	fUtxo := uint64(backingSize) * coin.FeeRate
-	return swapVal + fBase + fUtxo
+import (
+	"math"
+
+	"decred.org/dcrdex/dex"
+)
+
+// RequiredOrderFunds calculates the required amount needed to fulfill the swap
+// amount and pay transaction fees. swapVal is the total quantity needed to
+// fulfill an order. inputsSize is the size of the serialized inputs associated
+// with a set of UTXOs to be spent in the *first* swap txn. baseSwapSize is the
+// size of a standard swap with no inputs. The chained swap txns will be the
+// standard size as they will spend a previous swap's change output.
+// TODO: change the inputs when/if we eliminiate dex.Asset.SwapSize.
+func RequiredOrderFunds(swapVal, inputsSize, baseSwapSize uint64, coin *dex.Asset) uint64 {
+	maxSwaps := swapVal / coin.LotSize
+	if maxSwaps == 0 {
+		return math.MaxUint64 // caller messed up
+	}
+
+	baseBytes := maxSwaps * coin.SwapSize
+	// SwapSize already includes one input, replace the size of the first swap
+	// in the chain with the given size of the actual inputs + baseSwapSize.
+	firstSwapSize := inputsSize + baseSwapSize
+	totalBytes := baseBytes + firstSwapSize - coin.SwapSize // in this order because we are using unsigned integers
```

Comment:
I was hoping to chat about this.  I actually thinking `SwapSize` may not need to be a server-set field, and then we could have `RequiredOrderFunds` take both `baseSwapSize` and `swapSize`.  Although I understand there's no guarantee that the client and server share the same version of the `dex/networks` packages (or use it at all) where these values are defined.  Still, the consequences of not funding enough are just the inability to submit orders rather than failed swaps.  Hmm, if there's a way to avoid adding more data to the messages that would be better, but it looks like you're right.